sw/device/lib/base/print.c

```diff
@@ -0,0 +1,191 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/print.h"
+
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+// Standard format specifiers.
+static const char kPercent = '%';
+static const char kCharacter = 'c';
+static const char kString = 's';
+static const char kSignedDec1 = 'd';
+static const char kSignedDec2 = 'i';
+static const char kUnsignedOct = 'o';
+static const char kUnsignedHexLow = 'x';
+static const char kUnsignedHexHigh = 'X';
+static const char kUnsignedDec = 'u';
+static const char kCharsSoFar = 'n';
+static const char kPointer = 'p';
+
+// Verilog-style format specifiers.
+static const char kSvBinary = 'b';
+static const char kSvHexLow = 'h';
+static const char kSvHexHigh = 'H';
+
+static const char kDigitsLow[16] = "0123456789abcdef";
+static const char kDigitsHigh[16] = "0123456789ABCDEF";
+
+static const char kErrorNul[17] = "%<unexpected nul>";
+static const char kUnknownSpec[15] = "%<unknown spec>";
+
+static void blackhole(const char *buf, size_t len) {}
+
+buffer_sink_t base_stdout = &blackhole;
```

Comment:
I dislike using a global for this. `base_set_stdout`? `base_set_write_function`?

Reply:
Sg.